numpy/random/generator.pyx

```diff
@@ -3914,9 +3914,11 @@ cdef class Generator:
             return arr
 
         # Shuffle index array, dtype to ensure fast path
-        idx = np.arange(arr.shape[0], dtype=np.intp)
+        idx = np.arange(arr.shape[axis], dtype=np.intp)
```

Comment:
Should there be some input validation in case `axis` is a slice or something weird?

Reply:
Thanks for your reply! I added codes for error handling and updated the documentation :)

Reply:
I think what you want is `numpy.core.mulitarray.normalize_axis_index`, documented here https://github.com/numpy/numpy/blob/5a08e209863038b411bc506893093c5a4b8377c9/numpy/add_newdocs.py#L6739

Reply:
I didn't know this function! Thanks! Updated it.

Reply:
`normalize_axis_index` is all you need. You can remove the `isinstance` check and raising the error.

```python console
>>> import numpy
>>> numpy.__version__
'1.16.0'
>>> numpy.core.multiarray.normalize_axis_index(slice(1, 4, 2), 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: an integer is required (got type slice)
>>> numpy.core.multiarray.normalize_axis_index((slice(1, 4, 2),), 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: an integer is required (got type tuple)
>>> numpy.core.multiarray.normalize_axis_index((1, 2, None), 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: an integer is required (got type tuple)
>>> numpy.core.multiarray.normalize_axis_index(-1, 3)
2
>>> numpy.core.multiarray.normalize_axis_index(1, 3)
1
```

Reply:
I didn't remove `isinstance` because the input of `normalize_axis_index` is an `int`. But if it can also correctly raise the error, I can remove it.